adr/query_deduplication.tla

```diff
@@ -0,0 +1,485 @@
+            either {
+                \* expire one key
+                expiration_lock_1:
+                    cache[Key] := FALSE;
+                    call unlock(cache_lock);
+
+                expiration_done:
+                    goto finished;
+            }
+            or {
+                \*if let Some(value) = locked_cache.get(&key).cloned() { return value }
+                call unlock(cache_lock);
+                cache_hit:
+                    goto finished;
+            }
+        };
+
+
+    cache_miss:
+        \* we test at the end if all keys have been loaded
+        tested_keys := tested_keys \union {Key};
+
+        \* let mut locked_wait_map = self.wait_map.lock().await;
+        call lock(wait_map_lock);
+    
+    lock_3:
+        \* drop(locked_cache);
+        call unlock(cache_lock);
+    
+    lookup_wait_map:
+        \* match locked_wait_map.get_mut(&key) {
+        \*    Some(waiter) => {
+        if (Cardinality(wait_map[Key]) > 0)
+        {
+            wait_map_subscribe:
+                \* we register to reeive a notification when the value is available
+                \* let mut receiver = waiter.subscribe();
+                wait_map[Key] := wait_map[Key] \union {self};
+                \* drop(locked_wait_map);
+                call unlock(wait_map_lock);
+            
+            waiting_for_value:
+                \* let (recv_key, recv_value) = receiver.recv().await
+                await channels[self];
+                goto finished;
+    
+        }
+        else
+        {
+            wait_map_register:
+                \* the wait map should be emty for this key
+                assert wait_map[Key] = {};
+                \* we should not have a value yet
+                assert ~cache[Key];
+                
+                \* let (tx, _rx) = broadcast::channel(1);
+                \* locked_wait_map.insert(key.clone(), tx.clone());
+                wait_map[Key] := {self};
+                \* drop(locked_wait_map);
+                call unlock(wait_map_lock);
+                
+            fetching_value:
+                \* let value = self.resolver.retrieve(key.clone()).await;
+                skip;
+            
+            inserting_value_lock_cache:
+                \* let mut locked_cache = self.cached.lock().await;
+                call lock(cache_lock);
+                
+            inserting_value:
+                \* locked_cache.put(key.clone(), value.clone());
+                cache[Key] := TRUE;
+            
+                \* let mut locked_wait_map = self.wait_map.lock().await;
+                call lock(wait_map_lock);
+ 
+            remove_wait_map:
+                \* locked_wait_map.remove(&key);
+                \* locks are dropped right after last use
+
+                local_wait_map := wait_map[Key] \ { self };
+                wait_map[Key] := {};
+
+                call unlock(wait_map_lock);
+
+            unlock_1:
+                call unlock(cache_lock);
```

Comment:
this is the optimization I wanted to test in e8f32802881fc75f49ea521c609f3e711cf3582e: releasing the cache and wait map lock right after setting the cache.

I have been running this spec on a model with 5 tasks and 2 keys and it has found no issues. I will try an execution with more tasks, but It might reach the limit of what I can execute on my laptop